File: Regression/PLR.js
// Import required libraries
const fs = require('fs');
const csv = require('csv-parser');
const ml = require('ml-regression');
const PolynomialRegression = ml.PolynomialRegression;
const { plot } = require('nodeplotlib');

// Load and parse the dataset
const X = [];
const y = [];

fs.createReadStream('vehicles.csv')
  .pipe(csv())
  .on('data', (row) => {
    X.push(parseFloat(row['Battery Capacity (kWh)'])); // Battery Capacity
    y.push(parseFloat(row['Price ($1000)'])); // Price
  })
  .on('end', () => {
    // Split the dataset into training and testing sets (80-20 split)
    const splitIndex = Math.floor(0.8 * X.length);
    const X_train = X.slice(0, splitIndex);
    const y_train = y.slice(0, splitIndex);
    const X_test = X.slice(splitIndex);
    const y_test = y.slice(splitIndex);

    // Train the Polynomial Regression Model
    const degree = 4;
    const polyReg = new PolynomialRegression(X_train, y_train, degree);

    // Predictions
    const y_pred = X_test.map(x => polyReg.predict(x));

    // Evaluate the Model
    const mse = y_test.reduce((sum, yi, i) => sum + Math.pow(yi - y_pred[i], 2), 0) / y_test.length;
    const r2 = 1 - (mse / y_test.reduce((sum, yi) => sum + Math.pow(yi - (y_test.reduce((a, b) => a + b) / y_test.length), 2), 0));

    console.log(`Mean Squared Error (MSE): ${mse.toFixed(2)}`);
    console.log(`R-squared (R²): ${r2.toFixed(4)}`);

    // Visualization
    const sortedData = X.map((x, i) => ({ x, y: polyReg.predict(x) })).sort((a, b) => a.x - b.x);
    plot([
        { x: X, y: y, type: 'scatter', mode: 'markers', name: 'Actual Data', marker: { color: 'red' } },
        { x: sortedData.map(d => d.x), y: sortedData.map(d => d.y), type: 'scatter', mode: 'lines', name: 'Polynomial Fit', line: { color: 'blue' } }
    ], {
        title: 'Vehicle Price Prediction',
        xaxis: { title: 'Battery Capacity (kWh)' },
        yaxis: { title: 'Price ($1000)' }
    });
  });

File: Regression/PLR.py
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score

dataset = pd.read_csv('vehicles.csv')

X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, -1].values

poly_reg = PolynomialFeatures(degree=4)
X_poly = poly_reg.fit_transform(X)
X_train, X_test, y_train, y_test = train_test_split(X_poly, y, test_size=0.2, random_state=0)
regressor = LinearRegression()
regressor.fit(X_train, y_train)
y_pred = regressor.predict(X_test)

# evaluate the model
mse = mean_squared_error(y_test, y_pred)
print(f"Mean Squared Error (MSE): {mse:.2f}")

r2 = r2_score(y_test, y_pred)
print(f"R-squared (R²): {r2:.4f}")

# Scatter plot for actual data
plt.scatter(X, y, color='red')

# Sorted data for smooth curve
X_actual_sorted = np.sort(X, axis=0)  # Sort original features
X_poly_sorted = poly_reg.transform(X_actual_sorted)  # Apply transformation
y_poly_pred = regressor.predict(X_poly_sorted)  # Get predictions

# Polynomial regression curve
plt.plot(X_actual_sorted, y_poly_pred, color='blue')

# Labels
plt.title('Vehicle Price Prediction')
plt.xlabel('Battery Capacity (kWh)')
plt.ylabel('Price ($1000)')
plt.show()
